File: web/src/test/java/no/nav/sbl/dialogarena/filter/CORSFilter.java
package no.nav.sbl.dialogarena.filter;

import no.nav.sbl.dialogarena.sendsoknad.domain.util.ServiceUtils;

import javax.servlet.*;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.io.IOException;

public class CORSFilter implements Filter {

    @Override
    public void init(FilterConfig filterConfig) throws ServletException {}

    @Override
    public void doFilter(ServletRequest servletRequest, ServletResponse servletResponse, FilterChain filterChain) throws IOException, ServletException {
        String origin = "*";
        if (servletRequest instanceof  HttpServletRequest) {
            HttpServletRequest httpRequest = (HttpServletRequest) servletRequest;
            origin = httpRequest.getHeader("Origin");
        }

        HttpServletResponse httpResponse = (HttpServletResponse) servletResponse;
        httpResponse.setHeader("Access-Control-Allow-Origin", origin);
        httpResponse.setHeader("Access-Control-Allow-Headers", "Origin, Content-Type, Accept, X-XSRF-TOKEN");
        httpResponse.setHeader("Access-Control-Allow-Methods", "GET, POST, PUT, DELETE, OPTIONS");
        httpResponse.setHeader("Access-Control-Allow-Credentials", "true");
        filterChain.doFilter(servletRequest, httpResponse);
    }

    @Override
    public void destroy() {}
}

File: src/main/kotlin/no/nav/sosialhjelp/soknad/app/filter/CORSFilter.kt
package no.nav.sosialhjelp.soknad.app.filter

import jakarta.servlet.Filter
import jakarta.servlet.FilterChain
import jakarta.servlet.FilterConfig
import jakarta.servlet.ServletRequest
import jakarta.servlet.ServletResponse
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import no.nav.sosialhjelp.soknad.app.MiljoUtils
import org.springframework.stereotype.Component

@Component
class CORSFilter : Filter {
    override fun init(filterConfig: FilterConfig?) {
        super.init(filterConfig)
    }

    override fun doFilter(
        request: ServletRequest,
        response: ServletResponse,
        chain: FilterChain,
    ) {
        val httpResponse = response as HttpServletResponse
        val origin = if (request is HttpServletRequest) (request.getHeader("Origin")) else null

        if (MiljoUtils.isNonProduction() || ALLOWED_ORIGINS.contains(origin)) {
            httpResponse.setHeader("Access-Control-Allow-Origin", origin)
            httpResponse.setHeader(
                "Access-Control-Allow-Headers",
                "Origin, Content-Type, Accept, X-XSRF-TOKEN, Nav-Call-Id, Authorization, sentry-trace, baggage",
            )
            httpResponse.setHeader("Access-Control-Allow-Methods", "GET, POST, PUT, DELETE, OPTIONS")
            httpResponse.setHeader("Access-Control-Allow-Credentials", "true")
        }
        chain.doFilter(request, httpResponse)
    }

    companion object {
        private val ALLOWED_ORIGINS =
            listOf(
                "https://tjenester.nav.no",
                "https://www.nav.no",
            )
    }
}
